This notebook demonstrates how one might go about interfacing with a Zotero Group Library to harvest or pull its contents up for some particular use. In this case, we're looking specifically at how we might use a Zotero library as a source for a special collection of documents introduced into the xDD platform for NLP processing. The idea would be to initially baseline pull the full library and then run a scheduled proccess to check for updates and pull in new or changed metadata and documents.

Access will be through the Zotero API, and this notebook demonstrates that using the pyzotero abstraction on the REST API.

In [1]:
from pyzotero import zotero
from getpass import getpass
import pandas as pd
import json
import os

The Zotero API provides access to personal (user) and group libraries, with slightly different methods for each. Here we're only covering access to group libraries as this fits our use case. To instantiate an API connection with pyzotero, you need to provide the group library identifier and an API key. Library IDs are open and available to anyone. Here, we are interfacing with a library/collection of [NI 43-101 Reports](https://www.zotero.org/groups/4530692/usgs_ni_43-101_reports/library). You can see the library ID there in the URL: 4530692.

An API key has to be set up by an authorized user and will specify the type of access allowed and what library or libraries the key will access. In this case, we set up a read-only API key for the specific library, which should be the best practice for this kind of system.

In the following codeblock, we ask for these two parameters, hiding the API key as a secret. (Note: Contact the notebook author if you'd like to participate or set this up to work with your own specified library and API key.) In a production implementation of this idea, the Library ID and API Key parameters could be set up as environment variables for an application or some kind (e.g., Lambda process, etc.).

In [10]:
zotero_api_key = getpass("API Key")

inventory_item_key = "WSFB4RQE"
zotero_library_id = "4530692"

API Key ························


In [3]:
def zotero_library(library_id, api_key):
    # Establish connection to library with an ID and API Key
    zot = zotero.Zotero(
        library_id=library_id,
        library_type='group',
        api_key=api_key,
    )
    return zot

In [11]:
zot = zotero_library(zotero_library_id, zotero_api_key)

# Caching Library Metadata
For a Zotero Group Library that will be designated as an official GeoArchive repository, we will end up having a variety of services that periodically read information from the library and add services of perform functions.
* Backup metadata and file content to help ensure long-term viability
* Register identifiers/links to items via some form of "evergreen URL" capability
* Send items and content to additional value-added services (e.g., xDD AI engines)

To do this work, we will need to interact with collection and item metadata and file attachment content in various ways. After experimenting with various ways of working with the Zotero API and architecting a working system, it ended up making some sense to break up processing such that we periodically read from the library and cache some working files back to the library itself. A full baseline library item retrieval is pretty costly and time consuming via the Zotero API, even if we break up requests and operate them in parallel. The versioning mechanism and supported "since" queries provide a great way to periodically go after new and updated information once we have a baseline set.

I'm experimenting with setting up a top-level collection called "inventory" that will contain three JSON file attachments with all the essential data we need to operate against (collections, items, and tags). We can create a container app that will use a read/write API key to refresh these files on some scheduled basis. Downstream apps can then avoid needing to run their own costly API calls but instead check the files for updates compared against their own logs or other ways of checking for new information and then retrieve and operate using the file contents. Reading in a stored JSON file is immensely faster and less costly than running an API call for the same information.

In [58]:
def get_inventory_files(z, inventory_item):
    inventory_files = z.children(inventory_item, itemType="attachment")
    inventory_file_items = {}
    for file_type in ["items","collections","tags"]:
        inventory_file_items[file_type] = next((i for i in inventory_files if i["data"]["filename"] == f"{file_type}.json"), None)
        
    return inventory_file_items

def cache_library_tags(z, inventory_files, output_path="data"):
    tags = z.everything(z.tags())
    json.dump(tags, open(f"{output_path}/tags.json", w))

def baseline_items(z, output_path="data"):
    records = z.everything(z.items())
    json.dump(records, open(f"{output_path}/items.json", "w"))
    return len(records)

def baseline_collections(z, output_path="data"):
    records = z.everything(z.collections())
    json.dump(records, open(f"{output_path}/collections.json", "w"))
    return len(records)

def stash_tags(z, output_path="data"):
    records = z.everything(z.tags())
    json.dump(records, open(f"{output_path}/tags.json", "w"))
    return len(records)

def cache_inventory_data(
    z, 
    inventory_item, 
    output_path="data"
):
    existing_files = get_inventory_files(z, inventory_item)

    new_uploads = []
    for file_type in ["items","collections","tags"]:
        if os.path.exists(f"{output_path}/{file_type}.json"):
            if existing_files[file_type] is not None:
                z.delete_item(z.item(existing_files[file_type]["key"]))
            new_uploads.append(os.path.abspath(f"{output_path}/{file_type}.json"))
        
    if new_uploads:
        z.attachment_simple(new_uploads, parentid=inventory_item)
        print("Created new files in inventory", new_uploads)

def read_inventory_data(z, file_id):
    return z.file(file_id)

def load_inventory(z, inventory_item):
    existing_files = get_inventory_files(z, inventory_item)
    inventory = {}
    for file_type in ["items","collections"]:
        raw_data = read_inventory_data(z, existing_files[file_type]["key"])
        inventory[f"df_{file_type}"] = pd.DataFrame([i["data"] for i in raw_data])
        inventory[f"raw_{file_type}"] = raw_data
    
    raw_tags = read_inventory_data(z, existing_files["tags"]["key"])
    inventory["df_tags"] = pd.DataFrame([
            {
                "tag": i,
                "type": i.split(":")[0],
                "value": i.split(":")[-1]
            } for i in raw_tags if ":" in i
        ])
    inventory["raw_tags"] = raw_tags
    
    return inventory

def delete_from_cache(z, inventory_item, output_path="data"):
    inventory = load_inventory(z, inventory_item)
    deletions = z.deleted(since=0)

    deleted_items = [i for i in inventory["raw_items"] if i["key"] in deletions["items"]]
    deleted_collections = [i for i in inventory["raw_collections"] if i["key"] in deletions["collections"]]
    deleted_tags = [i for i in inventory["raw_tags"] if i in deletions["tags"]]
    
    print("ITEMS", len(deleted_items))
    print("COLLECTIONS", len(deleted_collections))
    print("TAGS", len(deleted_tags))
    
    if deleted_items:
        new_items = [i for i in inventory["raw_items"] if i["key"] not in deleted_items]
        

def update_stashes(z, inventory_item, output_path="data"):
    inventory = load_inventory(z, inventory_item)
    for k,v in inventory.items():
        if k.startswith("raw_"):
            print(f"STARTING NUMBER OF {k.split('_')[-1]}", len(v))

    last_updated_version_library = z.last_modified_version()
    print("LIBRARY LAST UPDATE VERSION", last_updated_version_library)

    items_last_updated = max(inventory["df_items"].version)
    print("CACHED ITEMS LAST UPDATE VERSION", items_last_updated)
    collections_last_updated = max(inventory["df_collections"].version)
    print("CACHED COLLECTIONS LAST UPDATE VERSION", collections_last_updated)

    deletion_effective_version = min([items_last_updated, collections_last_updated])
    print("EFFECTIVE VERSION FOR DELETIONS", deletion_effective_version)

    library_deletes = z.deleted(since=deletion_effective_version)
    print("ITEM DELETIONS RECORDED", len(library_deletes["items"]))
    print("COLLECTION DELETIONS RECORDED", len(library_deletes["collections"]))
    print("TAG DELETIONS RECORDED", len(library_deletes["tags"]))
    
    new_caches = {}
    for entity_type in ["items","collections"]:
        if library_deletes[entity_type]:
            new_caches[entity_type] = [i for i in inventory[f"raw_{entity_type}"] if i["key"] not in library_deletes[entity_type]]
        else:
            new_items_cache = inventory[f"raw_{entity_type}"]

    if library_deletes["tags"]:
        new_caches["tags"] = [i for i in inventory["raw_tags"] if i not in library_deletes["tags"]]
    else:
        new_caches["tags"] = inventory["raw_tags"]
        
    for k,v in new_caches.items():
        print(f"NEW NUMBER OF {k} AFTER DELETIONS", len(v))

    if last_updated_version_library > items_last_updated:
        new_items = z.everything(z.items(since=items_last_updated))
        new_items = [
            i for i in new_items if i["key"] != inventory_item
        ]
        new_items = [
            i for i in new_items if i["data"]["parentItem"] != inventory_item
        ]
        if new_items:
            print("NUMBER OF UPDATES/ADDITIONS", len(new_items))
            new_item_keys = [i["key"] for i in new_items]
            cached_item_keys = [i["key"] for i in new_caches["items"]]

            update_item_keys = [i for i in new_item_keys if i in cached_item_keys]
            print("ITEMS TO BE UPDATED", len(update_item_keys))

            add_item_keys = [i for i in new_item_keys if i not in cached_item_keys]
            print("ITEMS TO BE ADDED", len(add_item_keys))

            if update_item_keys:
                new_caches["items"] = [
                    i for i in new_caches["items"] 
                    if i["key"] not in update_item_keys
                ]

            new_caches["items"].extend(new_items)
            print("ITEMS TO BE CACHED", len(new_caches["items"]))
            json.dump(new_caches["items"], open(f"{output_path}/items.json", "w"))

    if last_updated_version_library > collections_last_updated:
        collection_updates = z.everything(z.collections(since=collections_last_updated))


#     if library_deletes["items"] or usable_item_updates:
#         updated_items = [
#             i for i in inventory["raw_items"] if i["key"] not in library_deletes["items"]
#         ]
#         if usable_item_updates:
#             updated_items = [
#                 i for i in updated_items if i["key"]
#                 not in [x["key"] for x in usable_item_updates]
#             ]
#             updated_items.extend(usable_item_updates)
            
#         json.dump(updated_items, open(f"{output_path}/items.json", "w"))
#         print("STASHED ITEMS", len(updated_items))

#     if library_deletes["collections"] or collection_updates:
#         updated_collections = [
#             i for i in inventory["raw_collections"] 
#             if i["key"] not in library_deletes["collections"]
#         ]
#         if collection_updates:
#             updated_collections = [
#                 i for i in updated_collections if i["key"]
#                 not in [x["key"] for x in collection_updates]
#             ]
#             updated_collections.extend(collection_updates)

#         json.dump(updated_collections, open(f"{output_path}/collections.json", "w"))
#         print("STASHED COLLECTIONS", len(updated_collections))
        
#     new_tags = z.tags(since=items_last_updated)
#     if library_deletes["tags"] or new_tags:
#         updated_tags = [
#             i for i in inventory["raw_tags"] 
#             if i not in library_deletes["tags"]
#         ]
#         updated_tags.extend(new_tags)
#         updated_tags = list(set(updated_tags))

#         json.dump(updated_tags, open(f"{output_path}/tags.json", "w"))
#         print("STASHED TAGS", len(updated_tags))


In [59]:
delete_from_cache(zot, inventory_item_key)

ITEMS 0
COLLECTIONS 0
TAGS 0


In [53]:
%%time
update_stashes(zot, inventory_item_key)

STARTING NUMBER OF items 30994
STARTING NUMBER OF collections 139
STARTING NUMBER OF tags 6484
LIBRARY LAST UPDATE VERSION 45895
CACHED ITEMS LAST UPDATE VERSION 45888
CACHED COLLECTIONS LAST UPDATE VERSION 45327
EFFECTIVE VERSION FOR DELETIONS 45327
ITEM DELETIONS RECORDED 3
COLLECTION DELETIONS RECORDED 1
TAG DELETIONS RECORDED 37
NEW NUMBER OF items AFTER DELETIONS 30994
NEW NUMBER OF collections AFTER DELETIONS 139
NEW NUMBER OF tags AFTER DELETIONS 6484
CPU times: user 1.44 s, sys: 207 ms, total: 1.64 s
Wall time: 8.13 s


In [57]:
zot.deleted(since=0)

{'collections': ['8AICMIB2', 'BCJ8XFGE', 'MGGTEK6C', 'NSMRK6EW', 'RW29JFBN'],
 'items': ['23N2K345',
  '23U33IWI',
  '33CXVW89',
  '367IE9DG',
  '3M4XR5JH',
  '3TXSKICC',
  '3VW58HDV',
  '4HDKB45X',
  '4PA7GJCN',
  '4RII95NP',
  '4RV4KS95',
  '4S3JI354',
  '54UH39PU',
  '63ZXI7WV',
  '6AT6J4ET',
  '6HXPHM58',
  '6ICXXIXA',
  '6PGHCTT2',
  '6Z33PWBT',
  '73HMNNE3',
  '77ZMEMR7',
  '7CMKIXM3',
  '7DVW3E7C',
  '7FJ8JP5X',
  '7RSMHR43',
  '7W3MWJFF',
  '86VP47TM',
  '9S43IDKN',
  'A833T4MX',
  'BDWPC7AR',
  'BG3JENCS',
  'BPSXHXIS',
  'BUFWD7DN',
  'DT4Q5PZU',
  'E9K3UA2U',
  'EANI37KV',
  'EP9U8QS7',
  'EX4K37BU',
  'F6QG8RMM',
  'FP9MHM5N',
  'I5Z354Z4',
  'J4DM736W',
  'J75ECMUH',
  'JCDRDHN2',
  'JD2ZQAAF',
  'JZTBNWXU',
  'M2XWP7PG',
  'M3TUVG9T',
  'NBM9HAHP',
  'P5UIXPIN',
  'PBGQ9JGX',
  'PDPNNTBK',
  'PIS2E8ZS',
  'PSAIQREQ',
  'R2TVZTZE',
  'R4QWDPSI',
  'RDRWEKBI',
  'RH9GUDWB',
  'RKVGFDK3',
  'S8S2M4NZ',
  'SBHV3DQM',
  'SDUNSR26',
  'SGQE29UQ',
  'SUIVZJ3F',
  'SWEMNANC',
  '

In [28]:
cache_inventory_data(zot, inventory_item_key)

Created new files in inventory ['/home/jovyan/experiments/data/items.json', '/home/jovyan/experiments/data/collections.json', '/home/jovyan/experiments/data/tags.json']


In [62]:
zot.item("7Z7SIR4E")

{'key': '7Z7SIR4E',
 'version': 45856,
 'library': {'type': 'group',
  'id': 4530692,
  'name': 'USGS NI 43-101 Reports',
  'links': {'alternate': {'href': 'https://www.zotero.org/groups/usgs_ni_43-101_reports',
    'type': 'text/html'}}},
 'links': {'self': {'href': 'https://api.zotero.org/groups/4530692/items/7Z7SIR4E',
   'type': 'application/json'},
  'alternate': {'href': 'https://www.zotero.org/groups/usgs_ni_43-101_reports/items/7Z7SIR4E',
   'type': 'text/html'},
  'enclosure': {'type': 'application/json',
   'href': 'https://api.zotero.org/groups/4530692/items/7Z7SIR4E/file/view',
   'title': 'inventory.json',
   'length': 48217844}},
 'meta': {'createdByUser': {'id': 1119084,
   'username': 'skybristol',
   'name': 'Sky Bristol',
   'links': {'alternate': {'href': 'https://www.zotero.org/skybristol',
     'type': 'text/html'}}},
  'numChildren': 0},
 'data': {'key': '7Z7SIR4E',
  'version': 45856,
  'itemType': 'attachment',
  'linkMode': 'imported_file',
  'title': 'inventor

In [54]:
inventory = load_inventory(zot, inventory_item_key)

In [55]:
inventory["df_items"]

key  version    itemType       linkMode  \
0      7Z7SIR4E    45853  attachment  imported_file   
1      TPHFX8G9    45843  attachment  imported_file   
2      DWECMZ8I    45841  attachment  imported_file   
3      DU4KNKKE    45839  attachment  imported_file   
4      UDPWD3HB    45837  attachment  imported_file   
...         ...      ...         ...            ...   
30989  E2FB73NU     6657      report            NaN   
30990  IFCEV3WF    45888      report            NaN   
30991  AXHWCJDI    45887      report            NaN   
30992  JXGEEQ33    45886      report            NaN   
30993  SK738P4V    45885      report            NaN   

                                                   title accessDate url note  \
0                                         inventory.json                       
1               4de78ee7-5acb-4f84-a6fa-b2b541e1432e.pdf                       
2               33ea7ddc-7cf2-4828-8de7-61c66f88adaf.pdf                       
3               bffc6efc-de53-4e55-9118-1a1a1999d151.pdf                       
4               68bb81a6-ce2a-46b8-8570-5da46027aab2.pdf                       
...                                                  ...        ...  ..  ...   
30989  NI 43-101 Technical Report for the Nord Tirek ...                 NaN   
30990  NI 43-101 Technical Report (1) for the F.W. Le...                 NaN   
30991  NI 43-101 Technical Report (1) for the F.W. Le...                 NaN   
30992  NI 43-101 Technical Report (1) for the F.W. Le...                 NaN   
30993  NI 43-101 Technical Report (1) for the F.W. Le...                 NaN   

            contentType charset  ...  date pages  language shortTitle archive  \
0      application/json          ...   NaN   NaN       NaN        NaN     NaN   
1       application/pdf          ...   NaN   NaN       NaN        NaN     NaN   
2       application/pdf          ...   NaN   NaN       NaN        NaN     NaN   
3       application/pdf          ...   NaN   NaN       NaN        NaN     NaN   
4       application/pdf          ...   NaN   NaN       NaN        NaN     NaN   
...                 ...     ...  ...   ...   ...       ...        ...     ...   
30989               NaN     NaN  ...  2009                                      
30990               NaN     NaN  ...  2004                                      
30991               NaN     NaN  ...  2007                                      
30992               NaN     NaN  ...  2004                                      
30993               NaN     NaN  ...  2007                                      

      archiveLocation libraryCatalog                            callNumber  \
0                 NaN            NaN                                   NaN   
1                 NaN            NaN                                   NaN   
2                 NaN            NaN                                   NaN   
3                 NaN            NaN                                   NaN   
4                 NaN            NaN                                   NaN   
...               ...            ...                                   ...   
30989                                 31a89683-673e-455b-bd23-b361d4d25c0e   
30990                                 fa7d0ccf-6642-4e12-9875-918f54491c72   
30991                                 7448b402-7756-45a6-ad24-1b8a336c6c23   
30992                                 ab856dd7-353b-4517-acbb-d0c9eb5efcac   
30993                                 874322da-02a4-4430-a3e5-a86e556f6219   

      rights    extra  
0        NaN      NaN  
1        NaN      NaN  
2        NaN      NaN  
3        NaN      NaN  
4        NaN      NaN  
...      ...      ...  
30989         05-2009  
30990         04-2004  
30991         05-2007  
30992         04-2004  
30993         05-2007  

[30994 rows x 36 columns]

# Collection-based Metadata
The organization scheme used in a Zotero Group Library could provide different types of value-added metadata about items. Collections are used like folders to provide a visual organizational scheme to library items, helping library users navigate through and work within a given library, particularly one that contains a lot of items.

In the case of the NI 43-101 Reports here, we replicated a scheme that has been used for years where these reports have been accumulating on network shared folders. It uses a hierarchy of folder names to indicate a region of the world and country as well as US States for mining projects that are the subject of the reports. Report items can be found at any of these levels with the immediate parent collection indicating the most specific geographic context available in the hierarchical structure. We may add a further level of structure with project names in future.

Since this information is contained in the organizational structure of the items and does not necessarily exist in individual item metadata (e.g., tags), we need to incorporate an "interpretation" of the organizational structure into our process of assembling useful metadata. In this case, the geographic context for a given report item may be useful in driving NLP processes. It can narrow the field on other, more specific place names we might work to recognize in full text or serve as a validation element in extracting point coordinates or other geospatial information.

Each process like this for reading a Zotero library and digesting its metadata is going to be different. The collection organization scheme is going to be different, introducing different types of information to the process. However, something like what we present here can be extended to essentially classify the collections in a given library as metadata elements of different kinds. For this case, we pull all collections and then use the hierarchical structure (via "parentCollection") to assign a "type" value of Region, Country, or US State to each collection. This essentially gives us an additional set of geographic context tags we can send along with our items for further use downstream.

When retrieving items (see below), we will get a list of collection keys that an item belongs to. We can also request items for a given collection. Any library item can belong to more than one collection. This essentially means we can use our collection metadata we assemble below to "backfill" additional value-added item metadata based on whatever significance the collections provide.

Organizing a relatively large library like the NI 43-101 Reports case into collections does present an opportunity to break up and parallelize processing of the library in various ways. The collection_items() function in pyzotero provides a way of retrieving the items of a given collection without delving into its subcollections, allowing us to work through a set of collections in a library, retrieving items and processing them in batches. 

Note: We did do some normalization work in our process of building out the Zotero library structure for the NI 43-101 Reports to ensure that each collection name can be resolved to official ISO sources for countries and US States. Regions represent continents and a few sub-continent but recognizable region names.

In [5]:
def library_collections(z):
    all_collections = z.everything(z.collections())

    collections = [
        {
            "type": "Region", 
            "key": i["key"], 
            "name": i["data"]["name"]
        } for i in all_collections if not i["data"]["parentCollection"]
    ]

    for r in collections:
        collections.extend([
            {
                "type": "Country", 
                "key": i["key"], 
                "name": i["data"]["name"],
                "parent": r["key"]
            } for i in all_collections if i["data"]["parentCollection"] == r["key"]
        ])

    us_collection = next((i for i in collections if i["name"] == "United States"), None)

    collections.extend([
        {
            "type": "US State", 
            "key": i["key"], 
            "name": i["data"]["name"],
            "parent": us_collection["key"]
        } for i in all_collections if i["data"]["parentCollection"] == us_collection["key"]
    ])
    
    return collections

In [6]:
collections = library_collections(zot)
pd.DataFrame(collections)

type       key             name    parent
0      Region  KVSV3898           Europe       NaN
1      Region  ZNCQ3V3A      Middle East       NaN
2      Region  BTJTAMND  Central America       NaN
3      Region  MQWZWWM2    North America       NaN
4      Region  BTZ7EAKX        Caribbean       NaN
..        ...       ...              ...       ...
163  US State  VGKU7ZGD          Georgia  AKV3WJ97
164  US State  5CQMNT5V          Florida  AKV3WJ97
165  US State  W8R6UH2U          Arizona  AKV3WJ97
166  US State  JK8JU3E2           Alaska  AKV3WJ97
167  US State  8JC7F475          Alabama  AKV3WJ97

[168 rows x 4 columns]

# Baselining Items
Via the Zotero API, everything in a given library is an item. This includes metadata principals for things like reports and articles, but file attachments are items along with collections and notes and anything else that is uniquely identified and accessed. The various item methods of pyzotero provide access to items and can be used in various ways to access and work through the contents of a library.

For our use case of retrieving documents for processing with the xDD tools, we will be focusing on pulling report type items for essential metadata, blending in collection item information (per the above discussion), to create xDD records. We'll also pull the file items attached to the report metadata items to get PDF document content for processing. Eventually, we will access and pull annotation items also attached to report items for value-added annotation that can be used in NLP training.

Again, each case may be a little bit different in terms of how a given library is organized and managed, resulting in some tweaks to this process for other circumstances. Hopefully, the basic patterns developed here can serve as a start to a more generalized process.

Baselining a given library for the purposes of creating an xDD collection or some similar purpose means working at a point in time to pull together all necessary item information translated into how the target system understands "item" for its context. 

After a library has been baselined, Zotero provides a useful versioning system that should be fairly convenient to use. Every change to any item results in a sequential version number being incremented. The version number is within the context of a given library (group or user). The last_modified_version() method of pyzotero can be used to simply retrieve the highest incremental version number across everything in the library. Usage patterns could include recording the high version number of the library at the point of sync and/or recording individual item versions and then using those within context for an API call, depending on how a workflow is composed. 

The API accepts a "since" parameter with version numbers that will only pull items changed after that version. This can be combined with other API search parameters to look for items changed within a given context (e.g., report type items changed after the last recorded version retrieved). In syncing items from a Zotero library to some other system, version numbers should be recorded and then used to go after newer items. The version number incremental system also applies for new items added to a library, so that if you ask for everything since the last version in a synced collection, you will get new items as well as updates.

Deleted items are a different issue that should also be taken into account. The pyzotero API provides a deleted method with the "since" parameter that returns lists of collections, items, and other things that were removed from the library. How these are dealt with in a target system like xDD is another consideration we'll have to work through, but the API method provides for pulling the deletions and deciding what actions to take.

The actual bibliographic metadata for items in a Zotero library is pretty straightforward and aligns well with various bibliographic metadata standards with information on title, authors, publishers, etc. The API allows for different standard forms of metadata to be requested, with the default being Zotero's own JSON format, so a harvester could be written to use something more standardized like bibtex. However, the Zotero JSON structure provides everything and is likely the easiest thing to build a comprehensize process upon.

The different itemTypes in Zotero control the schema for items, and the information elements in a given itemType schema are fixed. This introduces some challenges in recording all of the information we might want to record, but it produces a predictable structure with little room for interpretation on the downstream end. Pulling items into something like the xDD Digital Library means mapping the information elements to be encountered in Zotero libraries to the target schema.

In the xDD case, the target schema (viewable in the "fields" object [here](https://geodeepdive.org/api/articles)) is also very limited with a barebones set of identification metadata. So, our task of mapping from the barebones NI 43-101 Report metadata to the xDD "article" target is not difficult.

What will be more interesting is working through how we introduce what will essentially be value-added annotation into this process - additional metadata about the documents in the collection that we want to leverage in NLP processing. Geographic context and important name identifiers (projects, names of active mines, mineral and other commodities, etc.) are all important organizing principles that we are capturing anyway and recording through collections and tags. We may start introducing additional attachments on report items containing explicit geospatial data or more precise temporal information. We will be encouraging scientists and support staff to annotate PDF documents and use Zotero's capabilities to extract annotation (highlights and notes) as additional attachments in Markdown format. These processes will all introduce additional content that we want xDD and other systems to access and take advantage of.

Baselining the cache of a given library is a potentially resource-intensive process, with the biggest time delay in reading what we need from the Zotero API. The API is limited to 50 items in a given request as a throttling mechanism, with pyzotero offering a generator and a couple of other methods for pulling more items. After experimenting with several ways of breaking up the requests, I found that we don't end up buying very much through various parallelization strategies and it will be more simple and reliable to use the everything() wrapper in pyzotero to grab every item we need to operate on. The benchmark I ran comparing an everything request for 30,993 metadata and attachment items in the library vs. a Dask delayed process with 4 workers resulted in 21 minutes for the former and 13 minutes for the latter but a loss of some number of records that I couldn't explain but probably had to do with some API failure I didn't catch.

It does make sense, though, to get our items and then cache them as a file so that a) we can operate on that same file in different ways and b) we can store the file and use it to determine when the library contains something new for us to work on.

The three types of things we need from the Zotero library are collections (worked up previously), metadata items (reports and/or other itemTypes), and attachments. The pyzotero items() function, wrapped with everything(), will get us all of our items, or if we add the "since" parameter with a version number, we can get anything new.

Strategies for caching a library's raw information can vary depending on the platform where things are being implemented. If we're using a versioned file system like S3, we might opt to use that. Or we may come up with some file naming convention that works for our circumstance. In this rough workup, I'm simply using a local mounted file system and passing in a relative path. This will need to be tweaked to cover a broader range of use cases.

I still need to work up the following:

* Process for checking latest version in our cache against the last_modified_version from the library, getting new/changed items, and then either dealing with those separately or integrating them into our cache.
* Process for dealing with deleted items.
* Process for mapping Zotero metadata to a target schema like xDD articles.
* Process for managing file attachment content.

I split this up into three logical parts:
1) Get items for a given collection. This is the part that needs to make an API call to Zotero and may take the most resources to operate.
2) Process the metadata. This takes the raw item content and maps it to a profile, extracting the elements we need for that profile. It also handles blending in any additional metadata that might be useful but is outside the strict mapping to the target profile.
3) Get file attachment keys. The Zotero API offers a couple of different ways of retrieving files. They can be streamed through as bytes to some other process or sent directly to a file system.

In [21]:
def cache_library(z, output_path, output_file):
    items = z.everything(z.items())
    json.dump(items, open(f"{output_path}/{output_file}.json", "w"))
    return len(items)

def get_cached_library(library_path, library_file):
    items = json.load(open(f"{library_path}/{library_file}.json", "r"))
    return items

def cached_last_version(items):
    return max([i["version"] for i in items])

def z_collection_files(collection_items, file_type="application/pdf"):
    item_files = [
        {
            "item_key": i["key"],
            "attachment_key": i["links"]["attachment"]["href"].split("/")[-1]
        } for i in collection_items if "attachment" in i["links"]
        and i["links"]["attachment"]["attachmentType"] == file_type
    ]
    
    return item_files

In [9]:
africa = next((i for i in collections if i["name"] == "Africa"), None)

In [24]:
items = json.load(open("data/ni43101.json", "r"))

In [28]:
max([i["version"] for i in items])

45844

In [29]:
zot.last_modified_version()

45848

In [30]:
zot.items(since=45844)

[]

In [25]:
%%time
items = zot.everything(zot.items())

CPU times: user 5.22 s, sys: 340 ms, total: 5.56 s
Wall time: 21min 10s


In [30]:
next(i for i in items if i["data"]["itemType"] == "attachment")

{'key': 'TPHFX8G9',
 'version': 45843,
 'library': {'type': 'group',
  'id': 4530692,
  'name': 'USGS NI 43-101 Reports',
  'links': {'alternate': {'href': 'https://www.zotero.org/groups/usgs_ni_43-101_reports',
    'type': 'text/html'}}},
 'links': {'self': {'href': 'https://api.zotero.org/groups/4530692/items/TPHFX8G9',
   'type': 'application/json'},
  'alternate': {'href': 'https://www.zotero.org/groups/usgs_ni_43-101_reports/items/TPHFX8G9',
   'type': 'text/html'},
  'up': {'href': 'https://api.zotero.org/groups/4530692/items/HW37Q533',
   'type': 'application/json'},
  'enclosure': {'type': 'application/pdf',
   'href': 'https://api.zotero.org/groups/4530692/items/TPHFX8G9/file/view',
   'title': '4de78ee7-5acb-4f84-a6fa-b2b541e1432e.pdf',
   'length': 38806055}},
 'meta': {'createdByUser': {'id': 1119084,
   'username': 'skybristol',
   'name': 'Sky Bristol',
   'links': {'alternate': {'href': 'https://www.zotero.org/skybristol',
     'type': 'text/html'}}},
  'numChildren': 0}

In [59]:
reports[999]

{'key': 'QZM6W75P',
 'version': 6682,
 'library': {'type': 'group',
  'id': 4530692,
  'name': 'USGS NI 43-101 Reports',
  'links': {'alternate': {'href': 'https://www.zotero.org/groups/usgs_ni_43-101_reports',
    'type': 'text/html'}}},
 'links': {'self': {'href': 'https://api.zotero.org/groups/4530692/items/QZM6W75P',
   'type': 'application/json'},
  'alternate': {'href': 'https://www.zotero.org/groups/usgs_ni_43-101_reports/items/QZM6W75P',
   'type': 'text/html'},
  'attachment': {'href': 'https://api.zotero.org/groups/4530692/items/JS2XPAMU',
   'type': 'application/json',
   'attachmentType': 'application/pdf',
   'attachmentSize': 22635765}},
 'meta': {'createdByUser': {'id': 1119084,
   'username': 'skybristol',
   'name': 'Sky Bristol',
   'links': {'alternate': {'href': 'https://www.zotero.org/skybristol',
     'type': 'text/html'}}},
  'parsedDate': '2018',
  'numChildren': 1},
 'data': {'key': 'QZM6W75P',
  'version': 6682,
  'itemType': 'report',
  'title': 'NI 43-101 Te